[stdlib] FloatLiteral missing operators to conform Python float types

msaelices opened the pull request:
There are some missing operators in the `FloatLiteral` struct that we would need to implement to conform the Python `float` type:

* `__round__(ndigits)`
* `__mod__(rhs)`
* `__divmod__(rhs)`

This PR implements them and adds some tests to check the methods.

msaelices committed:
[stdlib] FloatLiteral missing operators to conform Python float types

Signed-off-by: Manuel Saelices <msaelices@gmail.com>

msaelices committed:
[stdlib] Some minor adjustments to the FloatLiteral.__divmod__() method and associated tests

Signed-off-by: Manuel Saelices <msaelices@gmail.com>

msaelices committed:
Merge branch 'nightly' into floatliteral-missing-operators-v2

Signed-off-by: Manuel Saelices <msaelices@gmail.com>

msaelices committed:
[stdlib] Minor dividend/quoting renaming

Signed-off-by: Manuel Saelices <msaelices@gmail.com>

msaelices committed:
[stdlib] Minor dividend/quoting renaming

Signed-off-by: Manuel Saelices <msaelices@gmail.com>

msaelices committed:
[stdlib] Do not execute a problematic test case, which IMO it's something related with the compiler (function unpacking) or the tuple type

Signed-off-by: Manuel Saelices <msaelices@gmail.com>

msaelices committed:
[stdlib] we need to comment all the problematic test code as the error is at compile time

Signed-off-by: Manuel Saelices <msaelices@gmail.com>

msaelices committed:
Merge branch 'nightly' into floatliteral-missing-operators-v2

laszlokindrat reviewed (commented):
This is awesome, thank you for pushing on this! I have a couple of questions and mostly small quibbles and suggestions, but we highly appreciate this improvement!

laszlokindrat reviewed (commented):
This is awesome, thank you for pushing on this! I have a couple questions and quibbles, but we really appreciate this contribution!

msaelices committed:
[stdlib] Minor fix in the FloatLiteral.__round__() docstring

Signed-off-by: Manuel Saelices <msaelices@gmail.com>

msaelices committed:
Merge branch 'nightly' into floatliteral-missing-operators-v2

msaelices committed:
[stdlib] Implement the Roundable trait in FloatLiteral

Signed-off-by: Manuel Saelices <msaelices@gmail.com>

msaelices committed:
[stdlib] Implement the Roundable trait in FloatLiteral

Signed-off-by: Manuel Saelices <msaelices@gmail.com>

msaelices committed:
[stdlib] Simplify test_mod() logic by using the % operator and just the float literals

Signed-off-by: Manuel Saelices <msaelices@gmail.com>

msaelices committed:
[stdlib] TODO about simplifying the FloatLiteral.__round__() logic when IntLiteral.__pow__ was implemented

Signed-off-by: Manuel Saelices <msaelices@gmail.com>

msaelices committed:
[stdlib] Delegate FloatLiteral.__mod__()'s logic to __divmod__()

Signed-off-by: Manuel Saelices <msaelices@gmail.com>

msaelices committed:
[stdlib] Remove debugging comments

Signed-off-by: Manuel Saelices <msaelices@gmail.com>

msaelices committed:
[stdlib] Add negative examples to FloatLiteral.__round__() with ndigits

Signed-off-by: Manuel Saelices <msaelices@gmail.com>

msaelices committed:
[stdlib] Fix the x.5 edge cases to have the same outcome as Python

Signed-off-by: Manuel Saelices <msaelices@gmail.com>

msaelices committed:
[stdlib] Test for the -2.5 and 2.5 cases

Signed-off-by: Manuel Saelices <msaelices@gmail.com>

msaelices committed:
[stdlib] Uncouple a little bit the __round__() logic with IntLiteral

Signed-off-by: Manuel Saelices <msaelices@gmail.com>

msaelices committed:
Merge branch 'nightly' into floatliteral-missing-operators-v2

msaelices committed:
[stdlib] Fix the FloatLiteral.__divmod__() tests

Signed-off-by: Manuel Saelices <msaelices@gmail.com>

Dan13llljws committed:
[stdlib] Provide safe `Variant` and `Optional` `take/get` methods

The safe version of `take/get` aborts if type checking fails.

MODULAR_ORIG_COMMIT_REV_ID: 240b627f31860a674e5878a1467e3c39feadce1b

avi-cenna committed:
[External] [docs] Correct typo in types.ipynb (#39888)

[External] [docs] Correct typo in types.ipynb

Correcting typo.

Co-authored-by: N <47500890+avi-cenna@users.noreply.github.com>
Closes modularml/mojo#2646
MODULAR_ORIG_COMMIT_REV_ID: aef3bb9dca96efdc6badc7b9369f14583ea2c71a

laszlokindrat committed:
[stdlib] Make `trunc` generic

This is done by introducing a `Truncable` trait. Implementations of
`__trunc__` are added for the core numeric types.

MODULAR_ORIG_COMMIT_REV_ID: 0b70cb6eaf05651ceb676eda40db6d364b0c9995

lattner committed:
[mojo-stdlib] Make `Tuple`'s constructor move its input elements. (#39904)

This changes `Tuple` to take its input pack as 'owned' and then
move from the pack into it storage.

This unearthed some bugs handling owned packs which were causing
multiple destructions, as a consequence of VariadicPack's ctor
moving off of `-> Self`.  Fix these bugs, make the variadics.mojo
more thorough, and make the IR pattern matching more "fail fast"
for when the IR changes again next time.

MODULAR_ORIG_COMMIT_REV_ID: b11da619445ce8330e7fef7b98e20860421ce1e6

msaelices committed:
Merge branch 'nightly' into floatliteral-missing-operators-v2

arthurevans committed:
[Docs] Fix a broken description of undeclared variables

Also, some general corrections and rewrites. (And one very trivial typo
fix in another file.)

MODULAR_ORIG_COMMIT_REV_ID: 70f8469144f5e17c9836fbd748a8b440707d53df

Dan13llljws committed:
[stdlib] Added `Variant.replace` (#39900)

[stdlib] Added `Variant.replace`

BEGING_PUBLIC
[stdlib] Added `Variant.replace`

The `replace` method serve the same purpose as `take` before, but
enforces placement of a new element immediately.

This also fix the problem where `Optional.take` does not actually
take out the value.

Test cases were updated accordingly
END_PUBLIC

---------

Co-authored-by: Connor Gray <code@connorgray.com>
MODULAR_ORIG_COMMIT_REV_ID: 0a35284f47a9a8fcf58c2a25e3288725ac426e37

abduld committed:
[Stdlib] Fix the abs implementation (#39100)

The abs function was evaluating both branches even if we can prove the
value is positive. Switch the conditions to avoid that.y

Co-authored-by: Abdul Dakkak <adakkak@m3.local>
MODULAR_ORIG_COMMIT_REV_ID: ee080ab8be7b8838d918734ee032fec34d12860b

Commit:
[External] [docs] Add constraint documentation for various `reduce`-like functions in simd.mojo (#39872)

Fixes #2237 for `reduce`-like functions in simd.mojo:
- Adds documentation for constraint on value of `size_out` not exceeding
width of input vector in `reduce_add`, `reduce_mul`, `reduce_max` and
`reduce_min`.

Signed-off-by: Brian-M-J <brianmj02@gmail.com>

Co-authored-by: Brian M Johnson <1DT21AI037@dsatm.edu.in>
Closes modularml/mojo#2461
MODULAR_ORIG_COMMIT_REV_ID: b151847dcc2e89c9b87fec873941bf8b20a534c7

peymanbr committed:
[External] [stdlib] Add unit tests for SIMD add, radd and iadd operations (#39931)

Add unit tests for the `__add__()`, `__radd__()`, and `__iadd__()`
methods.

Co-authored-by: Peyman Barazandeh <peymanb@gmail.com>
Closes modularml/mojo#2611
MODULAR_ORIG_COMMIT_REV_ID: b1616eedeb11b99a9e8f62df6f58cd708c587d78

laszlokindrat committed:
[stdlib] Re-enable `FloatLiteral.__trunc__` tests

These were accidentally disabled when the original patch landed.

MODULAR_ORIG_COMMIT_REV_ID: 0debdbd812763b50aefedc3aed34b51593fdd8c7

ConnorGray committed:
[stdlib] feature: Add initial string byte slicing

* Add String.as_bytes_slice() -> Span[Int8, _, _]
* Add StringLiteral.as_bytes_slice() -> Span[Int8, 0, ImmStaticLifetime]
* Add new ImmStaticLifetime and MutStaticLifetime helpers
* Add _byte_length() methods to String and StringLiteral
* Change Span.__getitem__() to return a Reference, making the doc
comment accurate
* Add Span.unsafe_ptr()
* Fix missing declaration of utils/span in CMakeLists.txt

MODULAR_ORIG_COMMIT_REV_ID: e7c7f512b0d968f1cbb1e2098f824b621046659a

ConnorGray committed:
[stdlib] Change `Reference.is_mutable` to `Bool` (from `i1`)

With the recent change to `Bool` to use an `i1` as its
representation, many of the errors holding up moving
`Reference.is_mutable` to `Bool` were resolved.

Co-authored-by: Chris Lattner <clattner@nondot.org>
MODULAR_ORIG_COMMIT_REV_ID: 980a32f9b676905284a58b454c3f5e25a7280b21

modularbot committed:
[stdlib] Bump compiler version to 2024.5.1515

msaelices committed:
[stdlib] FloatLiteral missing operators to conform Python float types

Signed-off-by: Manuel Saelices <msaelices@gmail.com>

laszlokindrat reviewed (approved):
Thanks again, it's great to have this fixed!
